src/lib.rs

```diff
@@ -80,24 +98,79 @@ pub mod __rt {
     /// the futures protocol.
     struct GenFuture<T>(T);
 
+    /// Small shim to translate from a generator to a stream.
+    struct GenStream<T> {
+        gen: T,
+        done: bool,
+    }
+
+    /// Uninhabited type to allow `await!` to work across both `async` and
+    /// `async_stream`.
+    pub enum Mu {}
+
     pub fn gen<T>(gen: T) -> impl MyFuture<T::Return>
-        where T: Generator<Yield = ()>,
+        where T: Generator<Yield = Async<Mu>>,
               T::Return: IsResult,
     {
         GenFuture(gen)
     }
 
+    pub fn gen_stream<T, U>(gen: T) -> impl MyStream<U, T::Return>
+        where T: Generator,
+              T::Yield: IsAsync,
+              <T::Yield as IsAsync>::Item: IsResult<Ok = U>,
+              T::Return: IsResult<
+                  Ok = (),
+                  Err = <<T::Yield as IsAsync>::Item as IsResult>::Err
+              >,
+    {
+        GenStream { gen, done: false }
+    }
+
     impl<T> Future for GenFuture<T>
-        where T: Generator<Yield = ()>,
+        where T: Generator<Yield = Async<Mu>>,
               T::Return: IsResult,
     {
         type Item = <T::Return as IsResult>::Ok;
         type Error = <T::Return as IsResult>::Err;
 
         fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
             match self.0.resume() {
-                GeneratorState::Yielded(()) => Ok(Async::NotReady),
-                GeneratorState::Complete(e) => e.into_result().map(Async::Ready),
+                GeneratorState::Yielded(Async::NotReady)
+                    => Ok(Async::NotReady),
+                GeneratorState::Yielded(Async::Ready(_))
+                    => unreachable!("Mu doesn't exist"),
+                GeneratorState::Complete(e)
+                    => e.into_result().map(Async::Ready),
+            }
+        }
+    }
+
+    impl<T> Stream for GenStream<T>
+        where T: Generator,
+              T::Yield: IsAsync,
+              <T::Yield as IsAsync>::Item: IsResult,
+              T::Return: IsResult<
+                  Ok = (),
+                  Err = <<T::Yield as IsAsync>::Item as IsResult>::Err
+              >,
+    {
+        type Item = <<T::Yield as IsAsync>::Item as IsResult>::Ok;
+        type Error = <<T::Yield as IsAsync>::Item as IsResult>::Err;
+
+        fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+            if self.done { return Ok(Async::Ready(None)) }
```

Comment:
I think it's ok to elide the boolean here, right?

Reply:
That's related to @Arnavion's comment: https://github.com/alexcrichton/futures-await/pull/32#issuecomment-334314136

I tried the test in https://github.com/alexcrichton/futures-await/pull/32/commits/e9d53a5e95d353a694ef168d725b8e7f219dfd1c to the changes before https://github.com/alexcrichton/futures-await/pull/32/commits/44fd004b71fe86aa5fcff975f4c78b6a9c26fbae and had it fail. Unless your conclusion from https://github.com/alexcrichton/futures-rs/issues/206#issuecomment-311594407 has changed then there needs to be the ability to track one more state change after the generator completes if it returns an error (and to support `?` it needs to retain the ability to return an error).

Reply:
Ah right yeah, makes sense